[tests] Reduced number of validations in tx_validationcache_tests

lucash-dev opened the pull request:
Following a suggestion in the comments, changed `ValidateCheckInputsForAllFlags` from testing all possible flag combinations to testing a random subset. Also created a new enum constant for the highest flag, so that this test doesn’t keep testing an incomplete subset in case a new flag is added.

Timing for `checkinputs_test`:
```
Before:   6.8s
After:    3.7s
----------------
Saved:    3.1s (45%)
```

This PR was split from #13050. Also see #10026.

lucash-dev committed:
Reduced number of validations in `tx_validationcache_tests` to keep the run time reasonable.

Following a suggestion in the comments, changed `ValidateCheckInputsForAllFlags` from testing all possible flag combinations to testing a random subset. Also created a new enum constant for the highest flag, so that this test doesn’t keep testing an incomplete subset in case a new flag is added.

laanwj commented on src/test/txvalidationcache_tests.cpp:
```diff
@@ -105,9 +105,11 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup)
+
+    for (uint32_t count=0; count < 10000; count++) {
+        // Randomly selects flag combinations
+        uint32_t test_flags = rand() % (HIGHEST_FLAG << 1);
```
we don't use `rand()` anywhere in the code, would be preferable to use a `FastRandomContext insecure_rand(true);` here, also for determinism - for reproducible it is undesirable for different test runs to test a different subset of flags

lucash-dev replied:
Fixed. Thanks for the explanation!

practicalswift commented on src/test/txvalidationcache_tests.cpp:
```diff
@@ -105,9 +105,13 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup)
+
+    for (uint32_t count=0; count < 10000; count++) {
+        // Randomly selects flag combinations
+        uint32_t test_flags = insecure_rand.randrange(HIGHEST_FLAG << 1);
```
Make this implicit conversion from `uint64_t` to `uint32_t` explicit :-)

lucash-dev replied:
Actually there's quite a lot of confusion of types around these types. `randrange` return a ` `uint64_t`, but `CheckInputs` accepts a `unsigned int`, while the flags are actually an enum which has an implementation-specific type. I guess the best is to change it here to `unsigned int` (and make the conversion explicit), and maybe open a separate PR for fixing these type inconsistencies.

adamjonas commented:
Concept ACK

I was looking at this before the conflict popped up and the savings on my machine was:
```
Before: 3.7s
After: 2.8s
------------
Saved 0.9s (24%).
```
My rebase is [here](https://github.com/adamjonas/bitcoin/tree/10-19-reduce-txvalidationcache-tests) for reference.

lucash-dev commented:
Thx @adamjonas.
Will be away for a few days, probably will rebase over the weekend.

lucash-dev commented:
@adamjonas rebased. thx

theStack reviewed (commented):
Concept ACK
Though on the current master branch enumerating through all possible script-verify-flags combinations is still feasible (running `tx_validationcache_tests` takes less than 2 seconds on my machine), the execution time for the validation loop would obviously increase exponentially with each script-verify-flag added.

lucash-dev commented:
> Concept ACK
> Though on the current master branch enumerating through all possible script-verify-flags combinations is still feasible (running `tx_validationcache_tests` takes less than 2 seconds on my machine), the execution time for the validation loop would obviously increase exponentially with each script-verify-flag added.

Hi, @theStack. Thank you for reviewing this PR. I agree with your comments and will implement them. Just a little bit busy, but hopefully will have some time for that in the next few weeks.

thanks!

lucash-dev commented:
hi @theStack. finally implemented the changes you suggested, also rebased -- and test are passing.

theStack reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/13533/commits/9364bf627eeb635a13378a436bc5c2707c63306a :beers:

adamjonas commented:
Hi @lucash-dev, pinging for a rebase. Mine is [here](https://github.com/bitcoin/bitcoin/compare/master...adamjonas:06-20-reduce-txvalidationcache-tests) for reference.

lucash-dev commented:
thx @adamjonas. rebased.

kallewoof commented on src/script/interpreter.h:
```diff
@@ -114,6 +114,12 @@ enum
     //
     SCRIPT_VERIFY_CONST_SCRIPTCODE = (1U << 16),
+
+
```
Nit: extraneous whitespace (1 empty line is usually enough).

kallewoof commented on src/script/interpreter.h:
```diff
@@ -114,6 +114,12 @@ enum
+    // Constants to point to the highest flag in use. Add new flags above this line.
+    //
+    SCRIPT_VERIFY_HIGHEST_FLAG_PLUS_ONE,
+    SCRIPT_VERIFY_HIGHEST_FLAG = SCRIPT_VERIFY_HIGHEST_FLAG_PLUS_ONE - 1
```
Suggest you call this `SCRIPT_VERIFY_END_MARKER`, and then use `< SCRIPT_VERIFY_END_MARKER` logic instead of this.

kallewoof commented on src/test/txvalidationcache_tests.cpp:
```diff
@@ -109,10 +109,15 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup)
+
+    FastRandomContext insecure_rand(true);
+
+    for (int count=0; count < 10000; count++) {
```
Style-nit: spacing, ++ before varname:
```C++
for (int count = 0; count < 10000; ++count) {
```

kallewoof commented on src/test/txvalidationcache_tests.cpp:
```diff
@@ -109,10 +109,15 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup)
         TxValidationState state;
+
+        // Randomly selects flag combinations
+        uint32_t test_flags = (uint32_t) insecure_rand.randrange(SCRIPT_VERIFY_HIGHEST_FLAG << 1);
```
Using the marker style, to get the same you'd `randrange((SCRIPT_VERIFY_END_MARKER - 1) << 1)` here.

However, I'm not sure why you are doing `<< 1` here (`1U << 17`), as the original code only goes up to `SCRIPT_VERIFY_END_MARKER - 1` (i.e. `(1U << 16)`). To retain the same, this would be `randrange(SCRIPT_VERIFY_END_MARKER - 1);`

lucash-dev replied:
Hi, @kalewoof. Thanks for reviewing.
The reason for that is that we want to possibly test all flag combinations (or a random subset thereof). The lowest value possible with all flags off is 0, the highest would be the sum of all constants. So if the highest constant is 1<<16, having the maximum value used for test be 1<<16 would preclude testing combinations of that flag being used. We need to test from 0 to `1111111111111111` (1<<17 -1). Because of the range being exclusive of the highest value, I pass `1<<17 = (SCRIPT_VERIFY_END_MARKER - 1) << 1`
Hope that clarifies.

lucash-dev replied:
The test currently only going to 1<<16 is probably a mistake caused by adding a flag but not updating the test.

lucash-dev replied:
See https://github.com/bitcoin/bitcoin/commit/9dabfe49c066301ef75bcfcb089fd308366127c4

lucash-dev replied:
Test created in https://github.com/bitcoin/bitcoin/commit/e3f9c05b966622146e090f2a01a913516ccb874a. Not updated after the new flag added in https://github.com/bitcoin/bitcoin/commit/9dabfe49c066301ef75bcfcb089fd308366127c4

kallewoof reviewed (commented):
Concept ACK; the change looks reasonable, and cutting down on redundant testing is definitely desirable.

lucash-dev commented:
Hi @fanquake
What’s the reason for closing it now?
I was meaning to implement the latest round of requested changes and it has concept acks.
Should I work on this, open a new one or just move on?
Just looking for clarification— don’t want to waste anyone’s time if maintainers don’t want this change. I know you have plenty on your plate.
Thanks!

fanquake commented:
Hi @lucash-dev. I closed this because of the length of time it's been open overall, the length of time it's been since comments were left and had been unaddressed, and after looking at your GitHub profile it seemed that you were no-longer active here (hence throwing it up for grabs). However if you are going to address the concerns, I'll re-open this.

lucash-dev commented:
Thanks @fanquake. Sorry for leaving this PR unattended for so long.

lucash-dev commented:
Thanks @fanquake. Sorry for leaving this PR unattended for so long.

lucash-dev commented:
Updated PR to fix @kallewoof's NIT's.

kallewoof commented on src/test/txvalidationcache_tests.cpp:
```diff
@@ -109,10 +109,15 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup)
         TxValidationState state;
+
+        // Randomly selects flag combinations
+        uint32_t test_flags = (uint32_t) insecure_rand.randrange((SCRIPT_VERIFY_END_MARKER - 1) << 1);
```
You want to swap the - and <<; right now you get `111...0`, i.e.
```suggestion
        uint32_t test_flags = (uint32_t) insecure_rand.randrange((SCRIPT_VERIFY_END_MARKER << 1) - 1);
```

lucash-dev replied:
Not really.
Right now the last flag is 1<<16.
Bc of the way the enum works my new constant becomes (1<<16)+1. That’s why I used the “...PLUS_ONE” constant before.
So the argument to the range is (((1<<16)+1)-1)<<1 = (1<<16)<<1 = 1<<17
Because the range is exclusive, the highest possible value is (1<<17)-1 = 1111111111111111b

kallewoof replied:
Ahh, I misread it as a mask for some reason.

lucash-dev commented:
Rebased.

@kallewoof would you mind ACK-ing since I addressed your comments (or otherwise make further comments)?

I think this PR has become (more) important after the new taproot flags were introduced -- since none of those flags are tested by this test case in `master`.

leonardojobim reviewed (commented):
tACK https://github.com/bitcoin/bitcoin/pull/13533/commits/c3e111a7daf5800026dda4455c737de0412528f1.
Tested on Ubuntu 20.04 on VMWAre.

Average tests execution time:
. Master branch: 1.02s
. This PR branch: 0.60s

These values have remained consistent across several runs.

kallewoof reviewed (commented):
ACK c3e111a7daf5800026dda4455c737de0412528f1

theStack reviewed (approved):
re-ACK c3e111a7daf5800026dda4455c737de0412528f1

lucash-dev commented:
Thanks everyone!

I know reviewing PRs is a lot of work, and this one wasn't a priority -- so I'm really glad to see it finally merged. 

I learned a lot from writing it and from your feedback!
